### PR TITLE
feat(auth): Add emulator support

### PR DESCRIPTION
**Pull request by cynthiajoan**
[replace this line]: # (Describe your changes in detail.)
***

[replace this line]: # (Describe your testing in detail.)
***

### Type of Change
Place an `x` the applicable box:
- [ ] Bug fix. Add the issue # below if applicable.
- [ ] New feature. A non-breaking change which adds functionality.
- [ ] Other, such as a build process or documentation change.
***

#### Notes
- Bug fixes and feature changes require an update to the `Release Notes` section of `release_build_files/readme.md`.
- Read the contribution guidelines [CONTRIBUTING.md](https://github.com/firebase/firebase-cpp-sdk/blob/main/CONTRIBUTING.md).
- Changes to the public API require an internal API review. If you'd like to help us make Firebase APIs better, please propose your change in a feature request so that we can discuss it together.

### CODE REVIEW

File: auth/src/desktop/rpcs/auth_request.cc [hunk 1]
```diff
@@ -77,5 +81,21 @@ AuthRequest::AuthRequest(::firebase::App& app, const char* schema,
   }
 }
 
+std::string AuthRequest::GetUrl() {
+  std::string emulator_url = ::firebase::auth::Auth::GetEmulatorUrl();
```
Comment (cynthiajoan) [inline auth/src/desktop/rpcs/auth_request.cc:85]: I'm trying to make this api desktop only, but cannot figure out a way to do that. Jon wdyt?

File: auth/src/android/auth_android.cc [hunk 2]
```diff
@@ -774,6 +775,18 @@ void Auth::SignOut() {
...
 
+void Auth::UseEmulator(const std::string host, uint32_t port) {
+  JNIEnv* env = Env(auth_data_);
+  jstring j_host = env->NewStringUTF(host.c_str());
+  env->CallVoidMethod(AuthImpl(auth_data_), auth::GetMethodId(auth::kUseEmulator), j_host, port);
+  env->DeleteLocalRef(j_host);
+  firebase::util::CheckAndClearJniExceptions(env);
+}
+
+std::string Auth::GetEmulatorUrl() {
```
Comment (jonsimantov) [inline auth/src/android/auth_android.cc:786]: Shouldn't we either save the host/port and return a URL here, or get it from the Android SDK?

File: auth/src/include/firebase/auth.h [hunk 3]
```diff
@@ -629,6 +630,16 @@ class Auth {
   /// to be called explicitly.
   void RemoveIdTokenListener(IdTokenListener* listener);
 #endif  // not SWIG
+
```
[the remark this answers is not in the document]
Comment (jonsimantov) [reply auth/src/include/firebase/auth.h:633]: Please run format_code.py!

File: auth/src/desktop/rpcs/delete_account_request.cc [hunk 4]
```diff
@@ -26,10 +26,8 @@ DeleteAccountRequest::DeleteAccountRequest(::firebase::App& app,
   FIREBASE_ASSERT_RETURN_VOID(api_key);
 
   const char api_host[] =
-      "https://www.googleapis.com/identitytoolkit/v3/relyingparty/"
       "deleteAccount?key=";
-  std::string url;
-  url.reserve(strlen(api_host) + strlen(api_key));
+  std::string url = GetUrl();
```
[the remark this answers is not in the document]
Comment (jonsimantov) [reply auth/src/desktop/rpcs/delete_account_request.cc:30]: Please fix, elsewhere too.

File: auth/src/ios/auth_ios.mm [hunk 5]
```diff
@@ -663,6 +663,17 @@ void SignInResultCallback(FIRAuthDataResult *_Nullable auth_result, NSError *_Nu
...
 
+void Auth::UseEmulator(const std::string host, uint32_t port) {
+  NSUInteger ns_port = port;
+  [AuthImpl(auth_data_) useEmulatorWithHost:@(host.c_str())
+                        port:ns_port
+  ];
+}
+
+std::string Auth::GetEmulatorUrl() {
+  return "";
```
Comment (jonsimantov) [inline auth/src/ios/auth_ios.mm:674]: Same comment as Android.

File: auth/src/android/auth_android.cc (hunk 2 above)
Comment (cynthiajoan) [reply auth/src/android/auth_android.cc]: Changed the way to use environment settings

File: auth/src/ios/auth_ios.mm (hunk 5 above)
Comment (cynthiajoan) [reply auth/src/ios/auth_ios.mm]: same with android, currently use environment settings, so don't need those apis

File: auth/src/desktop/rpcs/delete_account_request.cc (hunk 4 above)
Comment (cynthiajoan) [reply auth/src/desktop/rpcs/delete_account_request.cc:31]: done

File: auth/src/android/auth_android.cc [hunk 6]
```diff
@@ -185,6 +188,31 @@ void UpdateCurrentUser(AuthData* auth_data) {
   }
 }
 
+const char* const kEmulatorLocalHost = "10.0.2.2";
```
Comment (jonsimantov) [inline auth/src/android/auth_android.cc:191]: Just making sure, this is hard-coded for Android on purpose?

File: auth/src/desktop/rpcs/auth_request.cc [hunk 7]
```diff
@@ -77,5 +80,42 @@ AuthRequest::AuthRequest(::firebase::App& app, const char* schema,
   }
 }
 
+std::string AuthRequest::GetUrl() {
+  if (emulator_url.empty()) {
+    std::string url(kHttps);
+    url += kServerURL;
+    LogDebug("AuthRequest::GetUrl(Prod): %s", url.c_str());
```
Comment (jonsimantov) [inline auth/src/desktop/rpcs/auth_request.cc:87]: Keeping these debug statements in?

File: auth/src/ios/auth_ios.mm [hunk 8]
```diff
@@ -154,6 +157,24 @@ void UpdateCurrentUser(AuthData *auth_data) {
   SetUserImpl(auth_data, user);
 }
 
+void CheckEmulator(AuthData *auth_data) {
+  // Use emulator as long as this env variable is set, regardless its value.
+  if (std::getenv("USE_AUTH_EMULATOR") == nullptr) {
```
Comment (jonsimantov) [inline auth/src/ios/auth_ios.mm:162]: Just making sure, iOS also uses the environment variable?

File: release_build_files/readme.md [hunk 9]
```diff
@@ -629,6 +629,7 @@ code.
 ## Release Notes
 ### Upcoming Release
 -   Changes
+    - General: Add FirebaseApp.GetApps(), to return the list of firebase app.
```
Comment (jonsimantov) [inline release_build_files/readme.md:632]: nit: Add `FirebaseApp.GetApps(), to return the list of `firebase::App` instances.

File: release_build_files/readme.md [hunk 10]
```diff
@@ -629,6 +629,7 @@ code.
...
     - Firestore: Add support for disjunctions in queries (OR queries)
```
Comment (jonsimantov) [inline release_build_files/readme.md:633]: This code is already set up to use the emulator with environment variable set, are you sure we shouldn't add it to the readme already?

**Review by jonsimantov — APPROVED** [review]
Approved, but please take a look at my comments before merging.

File: auth/src/android/auth_android.cc (hunk 6 above)
Comment (cynthiajoan) [reply auth/src/android/auth_android.cc:191]: yes, I find some reference [here](https://cloud.google.com/firestore/docs/emulator#android_apple_platforms_and_web_sdks)

File: auth/src/ios/auth_ios.mm (hunk 8 above)
Comment (cynthiajoan) [reply auth/src/ios/auth_ios.mm:162]: Right we can set the environment variable through xcode running setup.

File: auth/src/desktop/rpcs/auth_request.cc (hunk 7 above)
Comment (cynthiajoan) [reply auth/src/desktop/rpcs/auth_request.cc:87]: Removed

File: release_build_files/readme.md (hunk 9 above)
Comment (cynthiajoan) [reply release_build_files/readme.md:633]: done

File: release_build_files/readme.md (hunk 10 above)
Comment (cynthiajoan) [reply release_build_files/readme.md:634]: as discussed, temporarily not adding it to release note

**Pull request merged**